File: src/main/java/com/ultreon/bubbles/entity/DamageableEntity.java
package com.ultreon.bubbles.entity;

import com.ultreon.bubbles.common.AttributeMap;
import com.ultreon.bubbles.common.gametype.AbstractGameType;
import com.ultreon.bubbles.effect.StatusEffectInstance;
import com.ultreon.bubbles.entity.attribute.Attribute;
import com.ultreon.bubbles.entity.damage.DamageSource;
import com.ultreon.bubbles.entity.modifier.Modifier;
import com.ultreon.bubbles.entity.player.PlayerEntity;
import com.ultreon.bubbles.entity.types.EntityType;
import com.ultreon.bubbles.event.EntityCollisionEvent;
import com.ultreon.bubbles.util.helpers.MathHelper;
import com.ultreon.hydro.common.ResourceEntry;
import com.ultreon.hydro.event.SubscribeEvent;
import com.ultreon.hydro.registry.Registry;
import org.bson.*;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Living Entity base class.
 * The class for all living entities, such as {@link PlayerEntity the player}.
 *
 * @see Entity
 */
@SuppressWarnings("unused")
public abstract class DamageableEntity extends Entity {
    protected float damageValue;
    protected double speed;
    protected double baseSpeed;
    private final List<Modifier> modifiers = new ArrayList<>();

    // Constructor.
    public DamageableEntity(EntityType<?> type, AbstractGameType gameType) {
        super(type, gameType);
    }

    // Properties
    public float getDamageValue() {
        return damageValue;
    }

    public void setDamageValue(float damageValue) {
        this.damageValue = MathHelper.clamp(damageValue, 0f, attributes.getBase(Attribute.MAX_DAMAGE));
    }

    public float getMaxDamageValue() {
        return attributes.getBase(Attribute.MAX_DAMAGE);
    }

    public double getSpeed() {
        return speed;
    }

    public void setSpeed(double speed) {
        this.speed = speed;
    }

    public double getBaseSpeed() {
        return baseSpeed;
    }

    @Override
    @SubscribeEvent
    public void onCollision(EntityCollisionEvent evtObj) {

    }

    // Attack & heal.

    /**
     * Attack!!!
     *
     * @param value The attack value.
     * @deprecated Use {@link #damage(double, DamageSource)} instead.
     */
    @Deprecated
    public void damage(double value) {
        if (attributes.getBase(Attribute.DEFENSE) == 0f) {
            this.destroy();
        }

        this.damageValue -= value / attributes.getBase(Attribute.DEFENSE);
        this.checkDamage();
    }

    /**
     * Attack!!!
     *
     * @param value  the attack value.
     * @param source the damage source.
     */
    @SuppressWarnings("unused")
    public void damage(double value, DamageSource source) {
        if (attributes.getBase(Attribute.DEFENSE) == 0f) {
            this.destroy();
        }

        this.damageValue -= value / attributes.getBase(Attribute.DEFENSE);
        this.checkDamage();
    }

    public void destroy() {
        this.damageValue = 0;
        checkDamage();
    }

    public void restoreDamage(float value) {
        this.damageValue += value;
        this.damageValue = MathHelper.clamp(damageValue, 0f, attributes.getBase(Attribute.MAX_DAMAGE));
    }

    protected void checkDamage() {
        if (this.damageValue <= 0) {
            this.delete();
        }
    }

    @Override
    public @NotNull BsonDocument getState() {
        BsonDocument state = super.getState();

        BsonArray bases = this.bases.write(new BsonArray());
        state.put("Bases", bases);

        BsonArray attributes = new BsonArray();
        attributes = this.attributes.write(attributes);
        state.put("Attributes", attributes);

        BsonArray effects = new BsonArray();
        for (StatusEffectInstance statusEffectInstance : activeEffects) {
            BsonDocument effectDocument = statusEffectInstance.write(new BsonDocument());
            effects.add(effectDocument);
        }
        state.put("Effects", effects);

        BsonDocument position = new BsonDocument();
        position.put("x", new BsonDouble(getX()));
        position.put("y", new BsonDouble(getY()));
        state.put("position", position);

        state.put("id", new BsonInt64(entityId));
        state.put("name", new BsonString(this.type.getRegistryName().toString()));

        state.put("damageValue", new BsonDouble(damageValue));

        return state;
    }

    @Override
    public void setState(BsonDocument state) {
        BsonArray attributes = state.getArray("Attributes");
        this.attributes = new AttributeMap().read(attributes);

        BsonArray bases = state.getArray("Bases");
        this.bases = new AttributeMap().read(bases);
//        this.baseSpeed = bases.getDouble("Speed").getValue();

        BsonDocument position = state.getDocument("position");
        this.x = (float) position.getDouble("x").getValue();
        this.y = (float) position.getDouble("y").getValue();

        // TODO: Update to locate entity class, from ResourceLocation using EntityType.
        ResourceEntry entityTypeKey = ResourceEntry.fromString(state.getString("Name").getValue());
//        this.type = EntityTypeRegistry.INSTANCE.get(entityTypeKey);
        this.type = Registry.getRegistry(EntityType.class).get(entityTypeKey);
    }

    public String toSimpleString() {
        return getRegistryName() + "@(" + Math.round(getX()) + "," + Math.round(getY()) + ")";
    }

    public String toAdvancedString() {
        BsonDocument nbt = getState();
        String data = nbt.toString();

        return getRegistryName() + "@" + data;
    }
}

File: src/main/kotlin/qtech/bubbles/entity/DamageableEntity.kt
package qtech.bubbles.entity

import qtech.bubbles.common.AttributeMap
import qtech.bubbles.common.entity.Attribute
import qtech.bubbles.common.entity.EntityDamageSource
import qtech.bubbles.common.entity.Modifier
import qtech.bubbles.common.gametype.AbstractGameMode
import qtech.bubbles.entity.types.EntityType
import qtech.bubbles.entity.player.PlayerEntity
import qtech.bubbles.event.CollisionEvent
import qtech.bubbles.event.SubscribeEvent
import qtech.bubbles.util.helpers.MathHelper
import com.qtech.preloader.PreGameLoader
import net.querz.nbt.tag.CompoundTag
import net.querz.nbt.tag.ListTag
import kotlin.math.roundToInt

/**
 * <h1>Living Entity base class.</h1>
 * The class for all living entities, such as [the player][PlayerEntity].
 *
 * @see Entity
 */
abstract class DamageableEntity (type: EntityType<*>, gameMode: AbstractGameMode) : Entity(type, gameMode) {
    open var damageValue: Float = 0f
        set(value) {
            field = MathHelper.clamp(value, 0f, attributeMap.get(Attribute.MAX_DAMAGE))
        }

    protected var bases = AttributeMap()

    var maxDamageValue: Float
        get() = attributeMap[Attribute.MAX_DAMAGE]
        set(value) {
            attributeMap.set(Attribute.MAX_DAMAGE, value)
        }

    var baseMaxDamageValue: Float
        get() = bases[Attribute.MAX_DAMAGE]
        protected set(value) {
            bases.set(Attribute.MAX_DAMAGE, value)
        }

    var attackDamage: Float
        get() = attributeMap[Attribute.ATTACK]
        set(value) {
            attributeMap.set(Attribute.ATTACK, value)
        }

    var baseAttackDamage: Float
        get() = bases[Attribute.ATTACK]
        protected set(value) {
            bases.set(Attribute.ATTACK, value)
        }

    var defenseValue: Float
        get() = attributeMap[Attribute.DEFENSE]
        set(value) {
            attributeMap.set(Attribute.DEFENSE, value)
        }

    var baseDefenseValue: Float
        get() = bases[Attribute.DEFENSE]
        protected set(value) {
            bases.set(Attribute.DEFENSE, value)
        }

    var speed: Float
        get() = attributeMap[Attribute.SPEED]
        set(value) {
            attributeMap.set(Attribute.SPEED, value)
        }

    var baseSpeed
        get() = bases[Attribute.SPEED]
        protected set(value) {
            bases.set(Attribute.SPEED, value)
        }

    var attributeMap = AttributeMap()
        protected set

    private val modifiers: List<Modifier> = ArrayList()

    @SubscribeEvent
    override fun onCollision(event: CollisionEvent) {
    }
    // Attack & heal.
    /**
     * <h1>Attack!!!</h1>
     *
     * @param value The attack value.
     */
    @Deprecated("Use {@link #damage(double, DamageSource)} instead.")
    fun damage(value: Double) {
        if (attributeMap.get(Attribute.DEFENSE) == 0f) {
            destroy()
        }
        damageValue -= value.toFloat()
        checkDamage()
    }

    /**
     * <h1>Attack!!!</h1>
     *
     * @param value  the attack value.
     * @param source the damage source.
     */
    open fun damage(value: Double, source: EntityDamageSource?) {
        if (attributeMap.get(Attribute.DEFENSE) == 0f) {
            PreGameLoader.LOGGER.warn("Defense is zero")
            destroy()
        }
        damageValue -= value.toFloat()
        checkDamage()
    }

    fun destroy() {
        damageValue = 0f
        checkDamage()
    }

    fun restoreDamage(value: Float) {
        damageValue += value
        damageValue = MathHelper.clamp(damageValue, 0f, attributeMap.get(Attribute.MAX_DAMAGE))
    }

    protected open fun checkDamage() {
        if (damageValue <= 0) {
            delete()
        }
    }

    override fun getState(): CompoundTag {
        val state = super.getState()
        val bases = bases.write(ListTag(CompoundTag::class.java))
        state.put("Bases", bases)
        var attributes: ListTag<CompoundTag?> =
            ListTag(CompoundTag::class.java)
        attributes = attributeMap.write(attributes)
        state.put("Attributes", attributes)
        val effects = ListTag(CompoundTag::class.java)
        for (effectInstance in activeEffects) {
            val effectDocument = effectInstance.write(CompoundTag())
            effects.add(effectDocument)
        }
        state.put("Effects", effects)
        val position = CompoundTag()
        position.putFloat("x", x)
        position.putFloat("y", y)
        state.put("position", position)
        state.putLong("id", entityId)
        state.putString("name", type.registryName.toString())
        state.putFloat("damageValue", damageValue)
        return state
    }

    override fun setState(state: CompoundTag) {
        val attributes = state.getListTag("Attributes", CompoundTag::class.java)
        attributeMap = AttributeMap().read(attributes)
        val bases = state.getListTag("Bases", CompoundTag::class.java)
        this.bases = AttributeMap().read(bases)
        val position = state.getCompoundTag("position")
        x = position.getFloat("x")
        y = position.getFloat("y")

        // TODO: Update to locate entity class, from ResourceLocation using EntityType.
//        val entityTypeKey = ResourceLocation.fromString(state.getString("Name"))
//        this.type = EntityTypeRegistry.INSTANCE.get(entityTypeKey);
//        type = Registry.getRegistry(EntityType::class.java)[entityTypeKey]!!
    }

    override fun toSimpleString(): String {
        return registryName.toString() + "@(" + x.roundToInt() + "," + y.roundToInt() + ")"
    }

    override fun toAdvancedString(): String {
        val nbt = getState()
        val data = nbt.toString()
        return "$registryName@$data"
    }
}